### PR TITLE
Oracle development

### PR DESCRIPTION
**Pull request by ShengguangXiao**
Oracle development

/kind feature

Fixes: https://github.com/cakedefi/defichain-private/issues/615

### CODE REVIEW

File: src/masternodes/rpc_oracles.cpp [hunk 1]
```diff
@@ -0,0 +1,1103 @@
...
+    CTransactionRef optAuthTx;
+
+    // check execution
+    {
+        LOCK(cs_main);
+        CCustomCSView mnview(*pcustomcsview); // don't write into actual DB
+        CCoinsViewCache coinview(&::ChainstateActive().CoinsTip());
+
+        auto &chain = pwallet->chain();
+        auto lock = chain.lock();
```
Comment (bvbfan) [inline src/masternodes/rpc_oracles.cpp:566]: chain lock is just `LOCK(cs_main)` what is doing some lines above, remove it.

File: src/masternodes/rpc_oracles.cpp [hunk 2]
```diff
@@ -0,0 +1,1103 @@
...
+    auto oracleRes = mnview.GetOracleData(oracleId);
+    if (!oracleRes.ok) {
+        throw JSONRPCError(RPC_DATABASE_ERROR, oracleRes.msg);
+    }
+
+    auto& oracle = *oracleRes.val;
+
+    UniValue result = OracleToJSON(mnview, oracle);
+
+    return result;
```
Comment (bvbfan) [inline src/masternodes/rpc_oracles.cpp:800]: Just `return` why extra variable and line?

**Review by bvbfan — CHANGES_REQUESTED** [review]

File: src/masternodes/mn_checks.cpp [hunk 3]
```diff
@@ -1614,35 +1615,36 @@ Res ApplySetOracleDataTx(CCustomCSView &mnview,
...
-    auto &&res = mnview.SetOracleData(msg.oracleId, msg.timestamp, msg.tokenPrices);
+    auto res = mnview.SetOracleData(msg.oracleId, msg.timestamp, msg.tokenPrices);
     if (!res.ok) {
         return Res::Err("SetRawPrice: %s", res.msg);
     }
 
     // TODO (IntegralTeam): ignore rpcInfo for now, implement getting tx info later
 
-    return Res::Ok(base);
+    return Res::Ok("Set oracle data");
```
Comment (bvbfan) [inline src/masternodes/mn_checks.cpp:1649]: I don't see reason to return string here, just return `Res::Ok()`

File: src/masternodes/mn_checks.cpp [hunk 4]
```diff
@@ -1614,35 +1615,36 @@ Res ApplySetOracleDataTx(CCustomCSView &mnview,
...
+    auto auth = oracleRes.val->oracleAddress;
     if (!skipAuth && !HasAuth(tx, coins, auth)) {
-        return Res::Err("%s: %s", base, "oracle authentication failed");
+        return Res::Err("%s: oracle authentication failed", __func__);
     }
 
-    auto &&res = mnview.SetOracleData(msg.oracleId, msg.timestamp, msg.tokenPrices);
+    auto res = mnview.SetOracleData(msg.oracleId, msg.timestamp, msg.tokenPrices);
     if (!res.ok) {
         return Res::Err("SetRawPrice: %s", res.msg);
```
Comment (bvbfan) [inline src/masternodes/mn_checks.cpp:1644]: `Res::Err("%s: %s", __func__, res.msg);`

**Review by bvbfan — COMMENTED** [review]
Overall is OK, rpc info is still missing, @ShengguangXiao did you tried as functinoallity?
Lint should be addressed as well.

**Comment by ShengguangXiao** [comment]
> Overall is OK, rpc info is still missing, @ShengguangXiao did you tried as functinoallity?
> Lint should be addressed as well.

Yes, i tried functionality, it works good.

**Comment by bvbfan** [comment]
@monstrobishi  it looks good, except `getcustomtx` lacks support of oracle txs.

**Comment by ShengguangXiao** [comment]
> @monstrobishi it looks good, except `getcustomtx` lacks support of oracle txs.

I already handle it. But the hardfork height should set to some height definition start with E?

**Comment by bvbfan** [comment]
@ShengguangXiao note that `rpcInfo` does nothing in these functions.

**Review by bvbfan — APPROVED** [review]

**Comment by ShengguangXiao** [comment]
@bvbfan @monstrobishi @uzyn I have tested this PR, it is ready to merge.

**Comment by bvbfan** [comment]
@surangap  you should add new hardfork check in few more places, take a look in old hardfork commits, then you should adjust the tests.

File: src/amount.h [hunk 5]
```diff
@@ -110,11 +110,7 @@ struct CTokenAmount { // simple std::pair is less informative
     CAmount nValue;
 
     std::string ToString() const {
-        const bool sign = nValue < 0;
-        const int64_t n_abs = (sign ? -nValue : nValue);
-        const int64_t quotient = n_abs / COIN;
-        const int64_t remainder = n_abs % COIN;
-        return strprintf("%s%d.%08d@%d", sign ? "-" : "", quotient, remainder, nTokenId.v);
+        return strprintf("%s@%d", GetDecimaleString(nValue), nTokenId.v);
```
Comment (monstrobishi) [inline src/amount.h:113]: Can you elaborate on this change?

File: src/masternodes/mn_checks.h [hunk 6]
```diff
@@ -30,44 +30,71 @@ enum CustomTxErrCodes : uint32_t {
...
+    SetOracleData         = 'y',
 };
 
-inline CustomTxType CustomTxCodeToType(unsigned char ch) {
-    constexpr const char txtypes[] = "CRTMNnpuslrUbBaGA";
-    if (memchr(txtypes, ch, sizeof(txtypes) - 1))
-        return static_cast<CustomTxType>(ch);
-    else
-        return CustomTxType::None;
+inline CustomTxType CustomTxCodeToType(uint8_t ch) {
```
Comment (monstrobishi) [inline src/masternodes/mn_checks.h:70]: will need to keep in mind this refactoring for the other PRs

Comment (bvbfan) [reply src/masternodes/mn_checks.h:70]: It was done by external dev.

File: src/amount.h (hunk 5 above)
Comment (bvbfan) [reply src/amount.h:113]: Duplicate code in `GetDecimaleString`

File: src/masternodes/mn_checks.h (hunk 6 above)
Comment (monstrobishi) [reply src/masternodes/mn_checks.h:70]: @bvbfan do you suggest we keep this change or revert to how it was before?

Comment (bvbfan) [reply src/masternodes/mn_checks.h:70]: We should add `ohty` in `txtypes` say `CRTMNnpuslrUbBaGAohty` so refactor is better, i think.

Comment (monstrobishi) [reply src/masternodes/mn_checks.h:70]: Yeah I think so too, it's cleaner, ok approving this

**Review by monstrobishi — APPROVED** [review]

**Pull request merged**